src/components/about-us/section-05/timeline.js

```diff
@@ -0,0 +1,176 @@
+    const frameId = window.requestAnimationFrame(this._loop)
+    this.setState({ frameId: frameId })    
+  }
+
+  _stopLoop = () => {
+    window.cancelAnimationFrame(this.state.frameId)
+  }
+
+  _setAutoScroll = () => {
+    if (!this.props.autoScrolling) return
+    if (!this._reachTheEnd()) {
+      this.setState({ timelineShiftY: this.state.timelineShiftY + 0.5 })
+      this._setYear()
+    } else {
+      this.setState({ returnToStart: true })
+      setTimeout(() => {
+        this.setState({ timelineShiftY: 0 })
+        this.setState({ returnToStart: false })
+      }, 500)
+    }
+  }
+
+  _reachTheEnd = () => {
+    return this.state.timelineShiftY >= this.props.childrenHeight
+  }
+
+  _onMouseDown = () => {
+    this.setState({ mouseDown: true })
+  }
+
+  _onMouseMove = (event) => {
+    if (!this.state.mouseDown) return
+    const clientY = event.clientY
+    this._timelineShifting(clientY)
+  }
+
+  _onMouseLeave = () => {
+    this.setState({ mouseDown: false })    
+  }
+
+  _onMouseUp = () => {
+    this.setState({ mouseDown: false })
+    this.setState({ prevClientY: null })
+  } 
+
+  _onMouseOver = () => {
+    this.props.stopAutoScroll()    
+  }
+
+  _onMouseOut = () => {
+    this.props.startAutoScroll()        
+  }
+
+  /**
+ * Set shifting value of time refering to the y-coordinate of pointer (mouse or touch) at this moment
+ * @param {number} coordY
+ */
+  _timelineShifting = (coordY) => {
+    if (!this.state.prevClientY) this.setState({ prevClientY: coordY })
+    if (this.state.prevClientY) {
+      const leftBound = 0
+      const shiftY = this.state.prevClientY - coordY
+      this.setState({ prevClientY: coordY })
+      if (this.state.timelineShiftY + shiftY >= leftBound && this.state.timelineShiftY + shiftY < this.props.childrenHeight) {
+        this.setState({ timelineShiftY: this.state.timelineShiftY + shiftY })
+        this._setYear()
+      }
+    } 
+  }
+
+  _setYear = () => {
+    const { yearContentHeight, getYear } = this.props
+    let currentYearIndex = 0
+    yearContentHeight.reduce((prev, curr, curIndex) => {
+      if (this.state.timelineShiftY > prev) {
+        currentYearIndex = curIndex
+        return prev + curr
+      }
+    })
+    getYear(currentYearIndex)
+  }
+
+  componentDidMount() {
+    this._startLoop()
+  }
+
+  componentWillUnmount() {
+    this._stopLoop()
+  }
+
+  render() {
+    return (
+      <div ref={elem => this.elem = elem}>
+        <Container returnToStart={this.state.returnToStart}>
+          <Scroller
+            shiftY={this.state.timelineShiftY}
+            onMouseDown={this._onMouseDown}
+            onMouseUp={this._onMouseUp}
+            onMouseMove={event => this._onMouseMove(event)}
+            onMouseLeave={this._onMouseLeave}
+            onMouseOver={this._onMouseOver}
+            onMouseOut={this._onMouseOut}
+          >
+            {this.props.children}
+          </Scroller>
+        </Container>  
+      </div>
+    )
+  }
+}
```

Comment:
Do we need to `setState` to trigger the componet to re-render?
Or just update the `transitionY` attribute directly(by `requestAnimationFrame`)?

Reply:
After changing the `timelineShiftY` to be a global variable and update the transitionY directly instead of using `setState`, the auto-scrolling animation is no longer smooth as before. Do I misunderstand something ?

Reply:
It's weird, may be we can discuss this issue in person.